### PR TITLE
[DataLoader] Introduce MapMapDataPipe functional datapipe

### LINKED ISSUE #57031: [DataLoader] Implement Map-style functional DataPipe
## 🚀 Feature
Add map-style Datapipe to support new DataLoader.

## Motivation
From the RFC #49440, you can find lots of issues with the previous DataLoader. So, we are working on a new design of DataLoader to support several features such as multithreading, and to prevent OOM. To support the new design of DataLoader, we have proposed the `IterDataPipe` for the iter-style DataPipe. But, in terms backward-compatibility, 
a list of functional `MapDataPipe` is also needed for users. The difference between iter-style and map-style can be found [here
](https://pytorch.org/docs/stable/data.html#dataset-types)

You can take reference from the implementation of [iter-style DataPipe](https://github.com/pytorch/pytorch/blob/6bbd8ba658d573956744dc68fed4614bd0fbd5d1/torch/utils/data/datapipes/iter/callable.py#L29-L82)

## Note
- path: torch/utils/data/datapipes/map/...
- `__len__` is required for map-style
- `__getitem__` is required for map-style
- Make sure DataPipe lazy-init and loading data lazily
- Make sure DataPipe is pickleable

## List
- [x] ConcatMapDataPipe
- [x] ZipMapDataPipe
- [x] BatchMapDataPipe
- [ ] UnbatchMapDataPipe (Likely do not want it)
- [x] MapMapDataPipe
- [ ] UnZipMapDataPipe (Likely do not want it)
- [x] ShuffleMapDataPipe
- [ ] FilterMapDataPipe (Likely do not want it)
- [ ] ForkMapDataPipe
- [ ] DemuxMapDataPipe (Likely do not want it, length not known in advance)
- [ ] MultiplexerMapDataPipe (Likely do not want it since the functionality doesn't make sense in Map-style? Open to suggestion if there is an actual use case)
- [ ] ... 

Here are the [complete list of `IterDataPipe`](https://pytorch.org/data/main/torchdata.datapipes.iter.html), some of which we may want to create Map-style version:

---
To avoid implementing every single feature twice (one in Iter-style, one in Map-style), our current [guidance](https://pytorch.org/data/main/torchdata.datapipes.map.html) is that users should use the built-in `IterDataPipe` for various functionalities, and convert it to MapDataPipe as needed using [`IterToMapConverter`](https://pytorch.org/data/main/generated/torchdata.datapipes.map.IterToMapConverter.html#torchdata.datapipes.map.IterToMapConverter) or .to_map_datapipe(). For that reason, I don't think we will need additional `MapDataPipe` for now.

Feedback and suggestions are welcome here.

cc @SsnL @VitalyFedyunin @ejguan @NivekT

### PR DESCRIPTION
**Pull request by MarcioPorto**
Summary:
As part of https://github.com/pytorch/pytorch/issues/57031, this PR adds the `MapMapDataPipe` functional datapipe for the `MapDataPipe` class.

Usage:
```
def fn(x):
    return x * 10

dp = CountingDataset(n=10)
dp.map(fn)
```

Differential Revision: D28394510

### CODE REVIEW

File: torch/utils/data/dataset.py [hunk 1]
```diff
@@ -40,6 +41,27 @@ def __add__(self, other: 'Dataset[T_co]') -> 'ConcatDataset[T_co]':
...
+
+    @classmethod
+    def register_datapipe_as_function(cls, function_name, cls_to_register):
+        if function_name in Dataset.functions:
+            raise Exception("Unable to add DataPipe function name {} as it is already taken".format(function_name))
+
+        def class_function(cls, source_dp, *args, **kwargs):
+            return cls(source_dp, *args, **kwargs)
+        function = functools.partial(class_function, cls_to_register)
+        Dataset.functions[function_name] = function
```
Comment (ejguan) [inline torch/utils/data/dataset.py:63]: We can get rid of the duplicate code for `IterableDataset` then.

**Review by ejguan — APPROVED** [review]
Thanks for adding the first `MapDataPipe`. LGTM. One comment below.

File: torch/utils/data/dataset.py (hunk 1 above)
Comment (VitalyFedyunin) [reply torch/utils/data/dataset.py:63]: We can not, this `Dataset. functions` have to be different from `IterableDataset.functions` unless you want carefull extract super class here and use it.

**Review by VitalyFedyunin — APPROVED** [review]

File: torch/utils/data/dataset.py (hunk 1 above)
Comment (ejguan) [reply torch/utils/data/dataset.py:63]: For two `classmethod` functions, we can change `Dataset` and `IterableDataset`to make it happened, since we are using decorator of `functional_datapipe` to dispatch. Am I right?

Comment (MarcioPorto) [reply torch/utils/data/dataset.py:63]: Yeah, that's correct. I can at least remove the two class methods using `cls`.
